Update for Guzzle 6

bakura10 opened the pull request:
This PR updates the client against Guzzle Services, which is based on Guzzle 6.

It includes several improvements:

* Service descriptor has been updated to new naming
* It includes a basic retry policy (up to 5 retries if a 429 is encountered)
* The ResourceIterator being removed in Guzzle 6, it has been replaced with a generator based pagination
* Pagination is now cursor based (using the since_id) instead of using page, which provide a more resilient pagination (in case new items are added during the pagination)
* The various `set` methods have been removed from the client, which makes it immutable. If you want to create a new client using different connection settings (such as a different shop domain or a different access token), you can now use the new `withConnectionOptions` which will create a new client.

@danizord can you please review? The removal of the `set` methods will likely cause some BC to our Common stuff but I think it can be easily updated :).

bakura10 commented on README.md:
````diff
@@ -51,11 +51,13 @@ $shopifyClient = new ShopifyClient([
 ```php
-$shopifyClient->setShopDomain('merchant.myshopify.com');
-$shopifyClient->setAccessToken('your_access_token');
+$newClient = $shopifyClient->withConnectionSettings([
````
We are doing that for instance in the Common: https://github.com/maestrooo/shopify-apps-common/blob/master/src/OAuth/ShopifyAuthorizationService.php#L89

Actually speaking about that I think that I'll need to remove the "validateSettings". In our case for instance we initiate the client, but we don't know the access token yet until we have loaded the info from the database.

What about removing completely the "withConnectionSettings", and asking always for a completetely created client? In our code, we may get ride of the the "shopifyClientPrototype" and just create the client using a new (would it be that bad ? Except it would be harder to test).

danizord replied:
>What about removing completely the "withConnectionSettings", and asking always for a completetely created client? In our code, we may get ride of the the "shopifyClientPrototype" and just create the client using a new

Could not agree more :+1:

bakura10 commented on src/ServiceDescription/Shopify-v1.php:
```diff
@@ -3723,5 +3816,13 @@
         ],
+    ],
+    'models' => [
+        'GetResponse' => [
```
Yeah... Actually in latest guzzle you need to set a "model". As I don't have the courage to write models for all response (Shopify payload are too complex) I just did as the doc and created a "default" model. But maybe GetResponse is not good :p. I'm goign to change it to "GenericModel".

bakura10 commented on src/ShopifyClient.php:
```diff
@@ -188,105 +198,61 @@
+ * @method \Generator getRecurringApplicationChargesIterator(array $commandArgs = [], array $iteratorArgs = []) {@command Shopify GetRecurringApplicationCharges}
+ * @method \Generator getSmartCollectionsIterator(array $commandArgs = [], array $iteratorArgs = []) {@command Shopify GetSmartCollections}
+ * @method \Generator getProductVariantsIterator(array $commandArgs = [], array $iteratorArgs = []) {@command Shopify GetProductVariants}
+ * @method \Generator getWebhooksIterator(array $commandArgs = [], array $iteratorArgs = []) {@command Shopify GetWebhooks}
```
Isn't it PHP 7.1 only?

bakura10 commented:
Added tests and fixed some issues after manual testing with the retrier (it happened that the exception was not set for 4xx errors so I need to rely only on the response).
